File: src/stack/stack_linked_list.py
from stack_interface import StackInterface
from src.list.node import Node

class StackLinkedList(StackInterface):
  """ stack implementation using a linked list """

  def __init__(self):
    """ create a new empty stack """
    self.head = None
    self.length = 0

  def isEmpty(self):
    """ verify is the stack is empty """
    return (self.length == 0)

  def push(self, cargo):
    """ add a new node to the stack """
    node = Node(cargo)

    if not self.isEmpty():
      node.next = self.head

    self.head = node
    self.length+=1

  def pop(self):
    """ remove and return the last added node from the stack """
    toBeRemoved = self.head
    self.head = self.head.next
    self.length-=1
    return toBeRemoved

File: ruby/lib/stack/stack_linked_list.rb
class StackLinkedList

  attr_accessor :head, :length

  def initialize
    @head = nil
    @length = 0
  end

  def empty?
    @length == 0
  end

  def push(cargo)
    node = Node.new(cargo)

    node.nextt = @head unless empty?

    @head = node
    @length+=1
  end

  def pop
    raise RangeError, 'the stack is empty!' if empty?
    removed_node = @head
    @head = @head.nextt
    @length-=1
    removed_node
  end
end
